### PR TITLE
Work from home

### LINKED ISSUE #59: Add possibility to request 'work from home' day
Main requirements:
1. user can take only one day in a week as 'work from home'
2. after user request bot should ask if work from home is approved by teem lead or manager
3. it should be not only a possibility to set date in format `(дд.мм)`, but it should be also aliases `сегодня` and `завтра`
4. approved day should be marked in google calendar
5. after user request `не работаю из дома` bot should reply `Ты хочешь, чтобы я вычеркнул дд.мм как день работы из дома? (да/нет)` or `Я еще не отмечал для тебя день работы из дома на этой неделе.` if user don't have approved work in home day
6. if user call off work from home day, bot should delete relevant mark from google calendar.

### PR DESCRIPTION
**Pull request by desa85**
close #59 
# Please make sure all below checkboxes in the Mandatory section have been checked before submitting your PR (also don't forget to pay attention to the Depending on Circumstances section)

## Mandatory

- [x] The commit message is an imperative and starts with a capital letter (for example, `Add something`, `Remove something`, `Fix something`, etc.).
- [x] I made sure I hadn't put a period at the end of the the commit message(s).

## Depending on Circumstances

Some of the items in the section become mandatory if you are a first-time contributor and/or your changes are relevant to the items.

- [ ] Adding a new parameter I didn't forget to reflect it in `README.md`.
- [ ] Modifying a list of constants or imports I didn't brake an alphabetical order.
- [ ] (for first-time contributors) One of the commits in the pull request adds me to the **Acknowledgements** section in `AUTHORS.md`.

### CODE REVIEW

File: src/viva.js [hunk 1]
```diff
@@ -249,39 +254,44 @@ module.exports = async (robot) => {
    * @param {String} end
    * @param {Object} user
   */
-  function addEventOfVacation (start, end, user) {
+  async function addEventGoogleCalendar (start, end, user, type) {
```
Comment (eugulixes) [inline src/viva.js]: Let's name it `addEventToCalendar`.

File: src/viva.js [hunk 2]
```diff
@@ -249,39 +254,44 @@ module.exports = async (robot) => {
...
+      })
+    }).catch(() => {
+      routines.rave(robot, 'An error occurred when attempting to add an event in the calendar.')
     })
+
+    return result.data.id
   }
 
-  function deletEventOfVacation (user) {
+  function deleteEventGoogleCalendar (eventId) {
```
Comment (eugulixes) [inline src/viva.js]: Let's name it `deleteEventFromCalendar`.

File: src/viva.js [hunk 3]
```diff
@@ -412,6 +422,90 @@ module.exports = async (robot) => {
...
+
+    if (message[5]) {
+      let day = moment().format('DD.MM')
+
+      if (moment(day, 'DD.MM').valueOf() <= moment(state.dateHomeToWork, 'DD.MM').valueOf()) {
+        msg.send(`Ты уже взял работу из дома на ${state.dateHomeToWork} . Можешь отменить работу из дома`)
+        return
+      }
+      if (aresidual > Math.abs(diff)) {
+        msg.send(`Ты уже взял работу из дома на этой неделе`)
```
Comment (eugulixes) [inline src/viva.js]: `Ты уже взял день работы из дома на этой неделе.`

File: src/viva.js [hunk 4]
```diff
@@ -412,6 +422,90 @@ module.exports = async (robot) => {
...
+        return
+      }
+    }
+
+    if (!message[1] && message[5]) {
```
Comment (eugulixes) [inline src/viva.js]: It's a bad practice. What do `message[1]` and `message[5]` mean? Nobody mustn't remember, so assign them to some variables.

File: src/viva.js [hunk 5]
```diff
@@ -8,6 +8,8 @@
 //
 // Commands:
 //   hubot хочу в отпуск - initiates a new leave request
+//   hubot работаю из дома - notes the status of work from home and add event in calendar
```
Comment (polina-popova) [inline src/viva.js]: Let's write `sets the status of work from home and adds event to the calendar`
And it is needed to add short version of the command to the help messages (`работаю из дома сегодня/завтра/ДД.ММ`)

File: src/viva.js [hunk 6]
```diff
@@ -8,6 +8,8 @@
...
+//   hubot не работаю из дома - removes status of work and deletes event
```
Comment (polina-popova) [inline src/viva.js]: Let's describe it as `removes work from home status and deletes event from the calendar`

File: src/viva.js [hunk 7]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+      day = moment()
+      dayOfWorkFromHome = moment(state.dateHomeToWork, 'DD.MM')
+      aresidual = +!!dayOfWorkFromHome.day() * 7 - dayOfWorkFromHome.day() + 1
+      diff = Math.floor(dayOfWorkFromHome.valueOf() / daysInDate - day / daysInDate) + 1
+    }
+
+    if (mainMessage) {
+      let day = moment().format('DD.MM')
+      if (moment(day, 'DD.MM').valueOf() <= moment(state.dateHomeToWork, 'DD.MM').valueOf()) {
+        msg.send(`Ты уже взял работу из дома на ${state.dateHomeToWork} . Можешь отменить работу из дома.`)
```
Comment (polina-popova) [inline src/viva.js]:
```
Ты уже работаешь из дома ${state.dateHomeToWork}.
Если хочешь все отменить, используй команду 'не работаю из дома' :wink:.
```

File: src/viva.js [hunk 8]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+      }
+      if (aresidual > Math.abs(diff)) {
+        msg.send(`Ты уже взял работу из дома на этой неделе.`)
+        return
+      }
+    }
+
+    if (!dateMessage && mainMessage) {
+      state.n = WAITING_DATE_STATE
+      msg.send('Окей, когда выходишь на работу из дома? \n `сегодня` `завтра` `DD.MM`')
```
Comment (polina-popova) [inline src/viva.js]: `Окей, когда ты хочешь работать из дома? (сегодня/завтра/DD.MM)`

File: src/viva.js [hunk 9]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+      } else if (dateMessage === 'завтра') {
+        day = moment()
+        date = day.date() + 1
+        month = day.month() + 1
+        dateOfWorkFromHome = moment(`${date}.${month}`, 'DD.MM').format('DD.MM')
+      } else {
+        if (moment(day, 'DD.MM').valueOf() <= moment(dateMessage, 'DD.MM').valueOf()) {
+          dateOfWorkFromHome = moment(dateMessage, 'DD.MM').format('DD.MM')
+        } else {
+          msg.send(`${dateMessage} уже прошло. Выбери дату что начинается с сегодняшнего дня`)
```
Comment (polina-popova) [inline src/viva.js]: `Но ${dateMessage} уже прошло :rolling_eyes:. Выбери дату позднее $(moment().format('D.M')).`

File: src/viva.js [hunk 10]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+    day = moment().format('DD.MM')
+
+    if (moment(day, 'DD.MM').valueOf() <= moment(state.dateHomeToWork, 'DD.MM').valueOf()) {
+      state.dateHomeToWork = moment(`${date}.${month}`, 'DD.MM').format('DD.MM')
+      state.dateRequested = null
+      if (GOOGLE_API && user.vivaLasVegas.homeWorkEventId) {
+        deleteEventFromCalendar(user.vivaLasVegas.homeWorkEventId)
+        delete user.vivaLasVegas.homeWorkEventId
+      }
+      msg.send('Я тебя понял. Убираю события с календаря.')
```
Comment (polina-popova) [inline src/viva.js]:
```
Я тебя понял, я тебя услышал. :ok_hand:
Убираю событие из календаря.
```

File: src/viva.js [hunk 11]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+    } else {
+      msg.send('Ты уже взял день работы из дома на этой неделе.')
```
Comment (polina-popova) [inline src/viva.js]: Why we need two separate messages for two almost similar situations?

File: src/viva.js [hunk 12]
```diff
@@ -528,6 +622,25 @@ module.exports = async (robot) => {
...
+        if (GOOGLE_API) {
+          const date = moment(`${state.dateHomeToWork}`, 'D.M')
+          const startDay = date.format('YYYY-MM-DD')
+          const endDay = date.add(1, 'days').format('YYYY-MM-DD')
+          const user = robot.brain.userForName(username)
+
+          const eventId = await addEventToCalendar(startDay, endDay, user, GOOGLE_EVENT_WORK_FROM_HOME)
+          user.vivaLasVegas.homeWorkEventId = eventId
+        }
+        msg.send(`Окей, уже пометил работу из дома в календаре.`)
```
Comment (polina-popova) [inline src/viva.js]: `Отлично, я уже отметил это за тебя в календаре. :sunglasses: `

File: src/viva.js [hunk 13]
```diff
@@ -528,6 +622,25 @@ module.exports = async (robot) => {
...
+      } else {
+        state.n = INIT_STATE
+        msg.send('Сперва согласуй все с руководителем, а пом повотори заявку заново.')
```
Comment (polina-popova) [inline src/viva.js]: `Сперва согласуй все с руководителем/лидом, а потом попробуй снова.`

File: src/viva.js [hunk 14]
```diff
@@ -412,6 +423,89 @@ module.exports = async (robot) => {
...
+        if (moment(day, 'DD.MM').valueOf() <= moment(dateMessage, 'DD.MM').valueOf()) {
+          dateOfWorkFromHome = moment(dateMessage, 'DD.MM').format('DD.MM')
+        } else {
+          msg.send(`${dateMessage} уже прошло. Выбери дату что начинается с сегодняшнего дня`)
+          return
+        }
+      }
+      state.dateRequested = dateOfWorkFromHome
+      state.n = WAITING_CONFIRMATION_STATE
+      msg.send(`${dateOfWorkFromHome}` + 'будешь работать дома. Согласован ли день из дома с руководителем/лидом? \n `да` `нет`')
```
Comment (polina-popova) [inline src/viva.js]: `${dateOfWorkFromHome}` + 
' ты хочешь работать дома. \n
Согласован ли день работы из дома с руководителем/лидом? (да/нет)'

File: src/viva.js [hunk 15]
```diff
@@ -249,39 +256,44 @@ module.exports = async (robot) => {
...
+        resource: event
+      }, (err, response) => {
+        if (err) {
+          reject(err)
+        } else {
+          resolve(response)
+        }
+      })
+    }).catch(() => {
+      routines.rave(robot, 'An error occurred when attempting to add an event in the calendar.')
```
Comment (eugulixes) [inline src/viva.js]: `...to the calendar.`

File: src/viva.js [hunk 16]
```diff
@@ -76,6 +76,46 @@ module.exports = async (robot) => {
     'Отправить текущую заявку в HR-отдел? (да/нет)'
   ])
 
+  class Stack {
+    constructor (date) {
+      this.firstDate = date[0]
+      this.secondDate = date[1]
+      this.today = moment().format('DD.MM')
```
Comment (eugulixes) [inline src/viva.js]: `OUTPUT_DATE_FORMAT` instead of `DD.MM`.

File: src/viva.js [hunk 17]
```diff
@@ -76,6 +76,46 @@ module.exports = async (robot) => {
...
+      this.comingDay = Stack.comingDay(this.secondDate, this.today)
+      this.thisWeek = Stack.thisWeek(this.secondDate, this.today)
+    }
+
+    get dates () {
+      return [this.firstDate, this.secondDate]
+    }
+
+    push (newDate) {
+      if (moment(this.secondDate, 'DD.MM').valueOf() < moment(this.today, 'DD.MM')) {
```
Comment (eugulixes) [inline src/viva.js]: `DATE_FORMAT` instead of `DD.MM` (in both cases).

File: src/viva.js [hunk 18]
```diff
@@ -76,6 +76,46 @@ module.exports = async (robot) => {
...
+
+    static thisWeek (date, today) {
+      let weekDay = moment(date).day()
+      let restOfDays = 7 - weekDay
+      let diff = moment(date, 'DD.MM').format('D') - moment(today, 'DD.MM').format('D')
+      return diff <= restOfDays && diff >= -weekDay
+    }
+
+    static comingDay (date, today) {
+      return moment(date, 'DD.MM').valueOf() > moment(today, 'DD.MM')
```
Comment (eugulixes) [inline src/viva.js]: `DATE_FORMAT` instead of `DD.MM` (in both cases).

File: src/viva.js [hunk 19]
```diff
@@ -76,6 +76,46 @@ module.exports = async (robot) => {
...
+
+    rollback () {
+      [this.firstDate, this.secondDate] = [this.firstDate, this.firstDate]
+      return this.dates
+    }
+
+    static thisWeek (date, today) {
+      let weekDay = moment(date).day()
+      let restOfDays = 7 - weekDay
+      let diff = moment(date, 'DD.MM').format('D') - moment(today, 'DD.MM').format('D')
```
Comment (eugulixes) [inline src/viva.js]: `DATE_FORMAT` instead of `DD.MM` (in both cases).

File: src/stack.js [hunk 20]
```diff
@@ -0,0 +1,108 @@
...
+    if (today.unix() > d.unix()) {
+      d.year(d.year() + 1)
+    }
+    return d.format(DATE_FORMAT)
+  }
+}
+
+exports.Stack = Stack
+
+console.log(moment('12', 'DD.MM.YYYY').format('DD.MM.YYYY'))
```
Comment (eugulixes) [inline src/stack.js]: Debug message? Should be removed from here.

File: src/viva.js [hunk 21]
```diff
@@ -8,6 +8,8 @@
 //
 // Commands:
 //   hubot хочу в отпуск - initiates a new leave request
+//   hubot работаю из дома - sets the status of work from home and adds event to the calendar
```
Comment (eugulixes) [inline src/viva.js]: `...adds the corresponding event...`

File: src/viva.js [hunk 22]
```diff
@@ -8,6 +8,8 @@
...
+//   hubot не работаю из дома - removes work from home status and deletes event from the calendar
```
Comment (eugulixes) [inline src/viva.js]: `...deletes the corresponding event...`

File: src/viva.js [hunk 23]
```diff
@@ -249,39 +267,53 @@ module.exports = async (robot) => {
...
+  function deleteEventFromCalendar (eventId) {
     GOOGLE_CALENDAR.events.delete({
       calendarId: GOOGLE_CALENDAR_ID,
-      eventId: user.vivaLasVegas.eventId
+      eventId: eventId
     }, (err) => {
       if (err) {
-        routines.rave(robot, 'An error occurred when attempting to delete an event in the calendar.')
-        return
+        routines.rave(robot, 'An error occurred when attempting to delete an event to the calendar.')
```
Comment (eugulixes) [inline src/viva.js]: Why was it changed to `...delete an event _to_ the calendar.` ?
Let's use the `from` preposition in this case to reflect the docstring.

File: src/stack.js [hunk 24]
```diff
@@ -0,0 +1,108 @@
...
+  canWork () {
+    let day = moment().format(DATE_FORMAT)
+    return moment(this[1], DATE_FORMAT).unix() < moment(day, DATE_FORMAT).unix()
+  }
+
+  /**
+   * Коды ответов:
+   * * 0 - OK
+   * * 1 - Invalid date
+   * * 2 - The `date` has passed
```
Comment (eugulixes) [inline src/stack.js]: The errors 3 and 4 are missing.

File: src/stack.js [hunk 25]
```diff
@@ -0,0 +1,108 @@
+// Добавить года
+// Изменить текст ошибки
```
Comment (eugulixes) [inline src/stack.js]: Let's remove it from here.

File: src/stack.js [hunk 26]
```diff
@@ -0,0 +1,108 @@
...
+      } else {
+        throw Error('Wrong data format.') // TODO
+      }
+    } else {
+      super(DEFAULT_DATE, DEFAULT_DATE)
+    }
+  }
+
+  /**
+   * Не находится ли [1] элемент в будущем
```
Comment (eugulixes) [inline src/stack.js]: Check if the 1st date is future.

File: src/stack.js [hunk 27]
```diff
@@ -0,0 +1,108 @@
...
+const DATE_FORMAT_SHORT = 'DD.MM'
+const DEFAULT_DATE = moment(0).format(DATE_FORMAT_SHORT)
+
+class Stack extends Array {
+  constructor (elements) {
+    if (elements) {
+      if (elements instanceof Array && elements.length === 2) {
+        super(...elements)
+      } else {
+        throw Error('Wrong data format.') // TODO
```
Comment (eugulixes) [inline src/stack.js]: `Incorrect date format.`

File: src/stack.js [hunk 28]
```diff
@@ -0,0 +1,108 @@
...
+    let day = moment().format(DATE_FORMAT)
+    return moment(this[1], DATE_FORMAT).unix() < moment(day, DATE_FORMAT).unix()
+  }
+
+  /**
+   * Коды ответов:
+   * * 0 - OK
+   * * 1 - Invalid date
+   * * 2 - The `date` has passed
+   * @param {String} date - date in "DD.MM" format
```
Comment (eugulixes) [inline src/stack.js]: `Date in 'DD.MM' format.`

File: src/stack.js [hunk 29]
```diff
@@ -0,0 +1,108 @@
...
+  /**
+   * Не находится ли [1] элемент в будущем
+   */
+  canWork () {
+    let day = moment().format(DATE_FORMAT)
+    return moment(this[1], DATE_FORMAT).unix() < moment(day, DATE_FORMAT).unix()
+  }
+
+  /**
+   * Коды ответов:
```
Comment (eugulixes) [inline src/stack.js]: `Return codes:`

File: src/stack.js [hunk 30]
```diff
@@ -0,0 +1,108 @@
...
+      }
+    }
+
+    // Everything's ok
+    return 0
+  }
+
+  /**
+   * Удаляет [0] элемент и вызывает .push()
+   * @param {String} el - Добавляемый элемент.
```
Comment (eugulixes) [inline src/stack.js]: `Element to be added.`

File: src/viva.js [hunk 31]
```diff
@@ -505,9 +600,33 @@ module.exports = async (robot) => {
...
+      const dateOfWorkFromHome = new Stack(state.dateOfWorkFromHome)
+      switch (dateOfWorkFromHome.checkDate(date)) {
+        case 1:
+          msg.send('Не валидная дата.')
+          return
+        case 2:
+          msg.send(`Но ${date} уже прошло :rolling_eyes:. Выбери дату позднее ${moment().add(-1, 'day').format(OUTPUT_DATE_FORMAT)}.`)
+          return
+        case 3:
+          msg.send('Нельзя брать работу из дома за 2 недели')
```
Comment (eugulixes) [inline src/viva.js]: `Нельзя запланировать день работы из дома больше, чем на две недели вперед.`

File: src/viva.js [hunk 32]
```diff
@@ -505,9 +600,33 @@ module.exports = async (robot) => {
...
+          return
+        case 4:
+          msg.send(`На этой неделе ты уже работал из дома, но ты можешь выбрать день на следующей.`)
```
Comment (eugulixes) [inline src/viva.js]: `Ты не можешь взять еще один день на этой неделе. Попробуй на следующей.`

**Pull request merged**